### PR TITLE
Add RepoDB with boltdb implementation

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/extensions/extension_search_disabled.go [hunk 1]
```diff
@@ -18,8 +19,8 @@ func EnableSearchExtension(config *config.Config, log log.Logger, rootDir string
 }
 
 // SetupSearchRoutes ...
-func SetupSearchRoutes(conf *config.Config, router *mux.Router,
-	storeController storage.StoreController, log log.Logger,
+func SetupSearchRoutes(config *config.Config, router *mux.Router, storeController storage.StoreController,
+	searchDB searchdb.SearchMetaDataDB, log log.Logger,
```
Comment (andaaron) [inline pkg/extensions/extension_search_disabled.go:23]: Why is the searchDB needed here?

File: pkg/extensions/extension_search.go [hunk 2]
```diff
@@ -65,17 +72,113 @@ func SetupSearchRoutes(config *config.Config, router *mux.Router, storeControlle
...
+			resConfig = search.GetResolverConfig(log, storeController, searchDB, false)
 		}
 
 		router.PathPrefix(constants.ExtSearchPrefix).Methods("OPTIONS", "GET", "POST").
 			Handler(gqlHandler.NewDefaultServer(gql_generated.NewExecutableSchema(resConfig)))
 	}
 }
 
+// reInitializeSearchDB resets the database and then repopulates with actual information.
+func reInitializeSearchDB(searchDB searchdb.SearchMetaDataDB, storeController storage.StoreController,
```
Comment (andaaron) [inline pkg/extensions/extension_search.go:92]: We need to sync on this.

File: pkg/storage/searchdb/boltdb_weapper_test.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package searchdb_test
```
Comment (bogdanbiv) [inline pkg/storage/searchdb/boltdb_weapper_test.go:1]: typo in filename weapper / wrapper

File: pkg/extensions/search/resolver.go [hunk 4]
```diff
@@ -205,163 +207,89 @@ func cleanQuerry(query string) string {
...
-		}
-
-		tagsInfo, err := olu.GetImageTagsWithTimestamp(repo)
-		if err != nil {
-			log.Error().Err(err).Msgf("can't get tags info for repo: %s", repo)
-
-			continue
-		}
+		var (
+			// map used for dedube if 2 images reference the same blob
```
Comment (andaaron) [inline pkg/extensions/search/resolver.go:223]: Minor typo.

File: pkg/extensions/search/resolver.go [hunk 5]
```diff
@@ -205,163 +207,89 @@ func cleanQuerry(query string) string {
...
-				}
-
-				if index := strings.Index(layer.Digest, searchName); index != -1 {
-					layers = append(layers, &gql_generated.LayerSummary{
-						Digest: &layer.Digest,
-						Size:   &layer.Size,
-						Score:  &index,
-					})
-				}
+			repo := repo
```
Comment (andaaron) [inline pkg/extensions/search/resolver.go:244]: Why is this needed?

File: pkg/extensions/search/resolver.go [hunk 6]
```diff
@@ -205,163 +207,89 @@ func cleanQuerry(query string) string {
 	return query
 }
 
-func globalSearch(searchResults map[string]map[string]string, searchName, searchTag string, olu common.OciLayoutUtils,
+func (r *queryResolver) globalSearch(searchResults map[string]map[string]string, searchName, searchTag string, olu common.OciLayoutUtils,
```
Comment (andaaron) [inline pkg/extensions/search/resolver.go:210]: Maybe we can remove the olu parameter?

**Review by andaaron — COMMENTED** [review]
Looks good overall.

On a separate note, I think we should sync with the rest of the team with regards to DB initialization, getting the parameters from the config file, and so on.

File: pkg/api/routes.go [hunk 7]
```diff
@@ -490,11 +493,157 @@ func (rh *RouteHandler) UpdateManifest(response http.ResponseWriter, request *ht
...
+		_, signedImageManifestDigest, _, err := storeController.DefaultStore.GetImageManifest(repoName,
+			manifestContent.Subject.Digest.String())
+
+		if err == nil && signedImageManifestDigest != "" {
+			return true, "notation", signedImageManifestDigest, nil
+		}
+	}
+
+	// check cosign
+	cosignTagRule := glob.MustCompile("sha256-*.sig")
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/routes.go:582]: They don't :(, they have "github.com/sigstore/cosign/pkg/oci/remote".SignatureTagSuffix, at least use this one.

Maybe we need a helper function for this, there are several places where we have this logic, can you consolidate them please?

File: pkg/storage/repodb/boltdb_wrapper.go [hunk 8]
```diff
@@ -0,0 +1,821 @@
...
+		}
+
+		return buck.Put([]byte(repo), repoMetaBlob)
+	})
+
+	return err
+}
+
+func (bdw BoltDBWrapper) DecrementRepoStars(repo string) error {
+	err := bdw.db.Update(func(tx *bolt.Tx) error {
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/storage/repodb/boltdb_wrapper.go:260]: From what I read from the documentation and from my tests calling Update, View or other methods that initiate a transaction on the DB object is thread safe and the database will remain consistent. 

"Creating transaction from the DB is thread safe" from the [documentation](https://github.com/etcd-io/bbolt#transactions).

File: pkg/storage/repodb/boltdb_wrapper.go [hunk 9]
```diff
@@ -0,0 +1,824 @@
...
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func (bdw BoltDBWrapper) AddManifestDependant(manifestDigest string, dependantDigest string) error {
```
Comment (andaaron) [inline pkg/storage/repodb/boltdb_wrapper.go:534]: We should not need this.

File: pkg/extensions/search/resolver.go [hunk 10]
```diff
@@ -238,170 +251,330 @@ func (r *queryResolver) repoListWithNewestImage(ctx context.Context, store stora
...
-				continue
-			}
+		var configContent ispec.Image
 
-			imageBlobManifest, err := olu.GetImageBlobManifest(repo, manifests[i].Digest)
-			if err != nil {
-				log.Error().Err(err).Msgf("can't read manifest for repo %s %s", repo, manifestTag)
+		err = json.Unmarshal(manifestMetaMap[manifestDigest].ConfigBlob, &configContent)
+		if err != nil {
+			continue
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/extensions/search/resolver.go:359]: I have updated the brances to return information about repo, tag and manifest digest of the image that failed.

File: pkg/extensions/search/resolver.go [hunk 11]
```diff
@@ -238,170 +251,330 @@ func (r *queryResolver) repoListWithNewestImage(ctx context.Context, store stora
...
+	for _, layer := range layers {
+		repoBlob2Size[layer.Digest.String()] = layer.Size
+		imgSize += layer.Size
+	}
+
+	return imgSize
+}
+
+func getImageLastUpdated(configContent ispec.Image) *time.Time {
+	var lastUpdated *time.Time
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/extensions/search/resolver.go:596]: Since it's an optional field I considered *nil* as a good default value. It tells us there is no info about the time when the image it was last updated

File: pkg/storage/repodb/boltdb_wrapper.go (hunk 9 above)
Comment (laurentiuNiculae) [reply pkg/storage/repodb/boltdb_wrapper.go]: I have removed it.

File: pkg/extensions/search/resolver.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/extensions/search/resolver.go:596]: ok

File: pkg/api/routes.go [hunk 12]
```diff
@@ -490,11 +493,157 @@ func (rh *RouteHandler) UpdateManifest(response http.ResponseWriter, request *ht
...
+	}
+
 	response.Header().Set("Location", fmt.Sprintf("/v2/%s/manifests/%s", name, digest))
 	response.Header().Set(constants.DistContentDigestKey, digest)
 	response.WriteHeader(http.StatusCreated)
 }
 
+func imageIsSignature(repoName string, manifestBlob []byte, manifestDigest, reference string,
+	storeController storage.StoreController,
+) (bool, string, string, error) {
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/routes.go:590]: you can call this GetSignatureInfoAndReference, and also you don't need to return bool value, you can assume that if signatureType, signedManifestDigest are not "", then is a signature.

**Comment by andaaron** [comment]
Close in favor of https://github.com/project-zot/zot/pull/830
